example/src/main/java/co/infinum/example/MainActivity.kt

```diff
@@ -133,33 +190,57 @@ class MainActivity : AppCompatActivity() {
         )
     }
 
-    private fun initGoldenEye() {
-        goldenEye = GoldenEye.Builder(this)
-            .setLogger(logger)
-            .setOnZoomChangedCallback { zoomView.text = "Zoom: ${it.toPercentage()}" }
-            .build()
-    }
-
     override fun onStart() {
         super.onStart()
-        if (goldenEye.availableCameras.isNotEmpty()) {
-            openCamera(goldenEye.availableCameras[0])
-        }
+        openCamera()
     }
 
     override fun onStop() {
         super.onStop()
-        goldenEye.release()
+        CameraX.unbindAll()
     }
 
-    private fun openCamera(cameraInfo: CameraInfo) {
+    private fun openCamera() {
         if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
-            goldenEye.open(textureView, cameraInfo, initCallback)
+            startCamera()
         } else {
             ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), 0x1)
         }
     }
 
+    private fun startCamera() {
+        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)
+
+        cameraProviderFuture.addListener(Runnable {
+            // Used to bind the lifecycle of cameras to the lifecycle owner
+            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()
+
+            // Preview
+            preview = Preview.Builder()
+                .build()
+            imageCapture = ImageCapture.Builder()
+                .build()
+            videoCapture = VideoCaptureConfig.Builder().build()
+
+            // Select camera lens
+            val cameraSelector = CameraSelector.Builder().requireLensFacing(lensFacing).build()
+
+            try {
+                // Unbind use cases before rebinding
+                cameraProvider.unbindAll()
+
+                // Bind use cases to camera
+                camera = cameraProvider.bindToLifecycle(
+                    this, cameraSelector, preview, imageCapture, videoCapture
+                )
```

Comment:
If I understood correctly, one key difference here is that you need to predefine use-cases and store them in a field. With GoldenEye, you can just open the camera and decide later what to do with it, although I suppose you're missing _some_ customization options.

Reply:
Yes, as far as I understood you need to define this use-cases, through them you actually set custom setting if you wish. If you want to change the settings in runtime you need to create new use cases, unbind old ones and create new ones. To me looks like GoldenEye does this internally because there you can update those settings in runtime

Reply:
It's not that it does so internally, but GoldenEye like Camera2 doesn't have the concept of "use-cases". When you initialize the camera, you can do whatever action you want later on.